### PR TITLE
Add server-side support for storing trip and place additions

### PR DESCRIPTION
**Pull request by shankari**
At a high level, we add two more data keys:
- trip_addition_input
- place_addition_input

The trip_addition_input is just a tripuserinput-type object. The place_addition_input is a new, parallel, placeuserinput type object

+ adding them in the timeseries and all the formatters

In addition:
- we add `confirmed_place` and `confirmed_untracked`, and match them with the respective user inputs
- create a confirmed timeline consisting of alternating `confirmed_*` objects
- create composite_trip objects that encapsulate confirmed trip-like and place-like objects

Add unit tests for everything

Includes changes from PRs:
https://github.com/shankari/e-mission-server/pull/4
https://github.com/shankari/e-mission-server/pull/5
https://github.com/shankari/e-mission-server/pull/6
https://github.com/shankari/e-mission-server/pull/7
https://github.com/shankari/e-mission-server/pull/8
https://github.com/shankari/e-mission-server/pull/9
https://github.com/shankari/e-mission-server/pull/10
https://github.com/shankari/e-mission-server/pull/11
https://github.com/shankari/e-mission-server/pull/12
https://github.com/shankari/e-mission-server/pull/13
https://github.com/shankari/e-mission-server/pull/15

And fixes for issues:
https://github.com/e-mission/e-mission-docs/issues/865
https://github.com/e-mission/e-mission-docs/issues/872
https://github.com/e-mission/e-mission-docs/issues/880

And is related to phone changes:
https://github.com/e-mission/e-mission-phone/pull/935
https://github.com/e-mission/e-mission-phone/pull/945

### CODE REVIEW

**Comment by shankari** [comment]
This PR includes the server side changes for https://github.com/e-mission/e-mission-docs/issues/840

File: emission/analysis/userinput/matcher.py [hunk 1]
```diff
@@ -27,29 +28,52 @@ def match_incoming_user_inputs(user_id):
...
+    multi_non_deleted_key_list = ["manual/trip_addition_input"]
+    input_key_list = single_most_recent_match_key_list + multi_non_deleted_key_list
     toMatchInputs = [ecwe.Entry(e) for e in ts.find_entries(input_key_list, time_query=timerange)]
-    logging.debug("Matching %d inputs to trips" % len(toMatchInputs))
+    logging.debug("Matching %d single inputs to trips" % len(toMatchInputs))
     lastInputProcessed = None
     if len(toMatchInputs) == 0:
         logging.debug("len(toMatchInputs) == 0, early return")
         return None
     for ui in toMatchInputs:
```
Comment (JGreenlee) [inline emission/analysis/userinput/matcher.py:40]: I have been having a hard time understanding how input matching works on phone vs server. I now understand why:

I see here that input matching on the server works by iterating through each user input (ui), and then ui is matched to the appropriate trip by esdt.get_trip_for_user_input_obj

This is fundamentally different on the phone: input matching works by iterating through each trip, and then finds the user inputs that correspond to that trip (with the functions getUserInputForTrip and getTripAdditionsForTrip in input-matcher.js)

If we want to unify the matching functions, and improve interoperability and maintainability, we would have to reconcile that fundamental difference

Comment (shankari) [reply emission/analysis/userinput/matcher.py:40]: @JGreenlee the server actually does both:
- in the common case, in which the user labels processed trips, we do indeed match the inputs to the trips
- however, for the uncommon case, in which the user labels draft trips, and we process both the trip information and the inputs at the same time, we match the trip to the inputs. Please see https://github.com/e-mission/e-mission-server/pull/895/commits/a7ee02c6e689f00368f88a219fd8d21c2da896ec

**Comment by shankari** [comment]
Test failures due to:

```
2023-04-16 17:51:57,331:ERROR:140704418879104:Error while creating confirmed objects, timestamp is unchanged
Traceback (most recent call last):
  File "/Users/kshankar/e-mission/gis_branch_tests/emission/analysis/userinput/matcher.py", line 94, in create_confirmed_objects
    confirmed_tl = create_and_link_timeline(ts, timeline, last_confirmed_place)
  File "/Users/kshankar/e-mission/gis_branch_tests/emission/analysis/userinput/matcher.py", line 119, in create_and_link_timeline
    timeline.first_place(), esda.CONFIRMED_PLACE_KEY, keys)
  File "/Users/kshankar/e-mission/gis_branch_tests/emission/analysis/userinput/matcher.py", line 181, in create_confirmed_entry
    get_user_input_dict(ts, tce, input_key_list)
  File "/Users/kshankar/e-mission/gis_branch_tests/emission/analysis/userinput/matcher.py", line 194, in get_user_input_dict
    matched_userinput = esdt.get_user_input_for_timeline_entry(ts, tct, ikey)
  File "/Users/kshankar/e-mission/gis_branch_tests/emission/storage/decorations/trip_queries.py", line 218, in get_user_input_for_timeline_entry
    return final_candidate(valid_user_input(ts, timeline_entry), potential_candidates)
  File "/Users/kshankar/e-mission/gis_branch_tests/emission/storage/decorations/trip_queries.py", line 164, in final_candidate
    extra_filtered_potential_candidates = list(filter(filter_fn, potential_candidate_objects))
  File "/Users/kshankar/e-mission/gis_branch_tests/emission/storage/decorations/trip_queries.py", line 159, in curried
    return valid_user_input_for_timeline_entry(ts, trip_obj, user_input)
  File "/Users/kshankar/e-mission/gis_branch_tests/emission/storage/decorations/trip_queries.py", line 125, in valid_user_input_for_timeline_entry
    (user_input.data.start_ts >= entry_start),
TypeError: '>=' not supported between instances of 'float' and 'NoneType'
```

**Comment by shankari** [comment]
Ok, so the error is because we now try to match against all timeline objects, not just trips, and the first place does not have an enter_ts.

```
2023-04-16 21:14:09,791:DEBUG:140704418879104:retrieved object
Entry({'_id': ObjectId('643cc783e0a6ae4f6c16dad7'),
'metadata': {'key': 'analysis/cleaned_place',
'data': {'source': 'DwellSegmentationTimeFilter',
'starting_trip': ObjectId('643cc781e0a6ae4f6c16da52'),
'exit_ts': 1466547704.0862284,
'exit_fmt_time': '2016-06-21T15:21:44.086228-07:00'
}) and added to id_map


2023-04-16 21:14:09,847:WARNING:140704418879104:Comparing user input bike (61af8dd3de1d3fd97454cc4f) of type manual/mode_confirm: 2016-06-21T15:21:44.086228-07:00 -> 2016-06-21T15:54:37.544000-07:00, trip of type 643cc783e0a6ae4f6c16dad7 2023-04-16T21:14:09.847257-07:00 (None) -> 2016-06-21T15:21:44.086228-07:00 (1466547704.0862284)
```

**Comment by shankari** [comment]
After fixing that, new error is 

```
======================================================================
FAIL: testCountLocalDateFullMissingLabelsMonth (__main__.TestMetrics)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "emission/tests/netTests/TestMetricsConfirmedTrips.py", line 270, in testCountLocalDateFullMissingLabelsMonth
    self.assertEqual(len(agg_met_result), 2)
AssertionError: 8 != 2

----------------------------------------------------------------------
Ran 1 test in 45.855s
```

The aggregate calls are in fact

```
[[ModeStatTimeSummary({'ts': 1451635200, 'local_dt': LocalDate({'year': 2016, 'month': 1, 'timezone': 'America/Los_Angeles'}), 'fmt_time': '2016-01', 'nUsers': 9, 'label_unlabeled': 60}),
ModeStatTimeSummary({'ts': 1454313600, 'local_dt': LocalDate({'year': 2016, 'month': 2, 'timezone': 'America/Los_Angeles'}), 'fmt_time': '2016-02', 'nUsers': 12, 'label_unlabeled': 84}),
ModeStatTimeSummary({'ts': 1464764400, 'local_dt': LocalDate({'year': 2016, 'month': 6, 'timezone': 'America/Los_Angeles'}), 'fmt_time': '2016-06', 'nUsers': 28, 'label_bike': 18, 'label_shared_ride': 18, 'label_unlabeled': 80, 'label_walk': 36}),
ModeStatTimeSummary({'ts': 1467356400, 'local_dt': LocalDate({'year': 2016, 'month': 7, 'timezone': 'America/Los_Angeles'}), 'fmt_time': '2016-07', 'nUsers': 12, 'label_unlabeled': 117}),
ModeStatTimeSummary({'ts': 1470045600, 'local_dt': LocalDate({'year': 2016, 'month': 8, 'timezone': 'Pacific/Honolulu'}), 'fmt_time': '2016-08', 'nUsers': 40, 'label_unlabeled': 272}),
ModeStatTimeSummary({'ts': 1472713200, 'local_dt': LocalDate({'year': 2016, 'month': 9, 'timezone': 'America/Los_Angeles'}), 'fmt_time': '2016-09', 'nUsers': 3, 'label_unlabeled': 18}),
ModeStatTimeSummary({'ts': 1475305200, 'local_dt': LocalDate({'year': 2016, 'month': 10, 'timezone': 'America/Los_Angeles'}), 'fmt_time': '2016-10', 'nUsers': 3, 'label_unlabeled': 18}),
ModeStatTimeSummary({'ts': 1480579200, 'local_dt': LocalDate({'year': 2016, 'month': 12, 'timezone': 'America/Los_Angeles'}), 'fmt_time': '2016-12', 'nUsers': 3, 'label_unlabeled': 12})]]
```

I wonder if this is because of other entries sitting around in the database since it is the aggregate test that is failing.
Let's delete the database and retry.

**Comment by shankari** [comment]
Yup! Test now passes.

```
----------------------------------------------------------------------
Ran 1 test in 38.622s

OK
```

**Comment by shankari** [comment]
Final failure

```
AssertionError: Lists differ: [1470[53 chars]470353892.225392, 1470354375.763474, 147035561[59 chars]4782] != [1470[53 chars]470354036.959, 1470354386.5618007, 1470355612.[57 chars]4782]

First differing element 3:
1470353892.225392
1470354036.959

  [1470341031.235,
   1470343292.1363852,
   1470352238.7396843,
-  1470353892.225392,
-  1470354375.763474,
+  1470354036.959,
+  1470354386.5618007,
   1470355612.592,
   1470356315.8404644,
   1470357578.288795,
   1470364485.744782]

----------------------------------------------------------------------
Ran 328 tests in 795.222s

FAILED (failures=1)
```

**Comment by shankari** [comment]
Difference:

```
1470353892.225392 -> 1470354036.959
2016-08-04T23:38:12.225392+00:00 -> 2016-08-04T23:40:36.959000+00:00


1470354375.763474 -> 1470354386.5618007
2016-08-04T23:46:15.763474+00:00 -> 2016-08-04T23:46:26.561801+00:00
```

Changes are very subtle, basically off by a few minutes/seconds. I wonder if the original values were incorrect to begin with?

**Comment by shankari** [comment]
Running it step by step manually in a REPL, I get the same values

**Comment by shankari** [comment]
So this:
- works on my laptop with the old environment (conda 4.8.3)
- works on my laptop with the new environment (conda 23.1.0)
- works for the test real examples on the github actions environment (this is the only failure, and reading the expected values confirms that they are the same)
- fails for this test only on the github actions environment

I am going to try setting the random seed and see if it works.
If not, I am going to merge this to get the heavy lifting done, and then focus on enabling only this one test

**Comment by shankari** [comment]
Setting the random seed does not fix it.
Commenting it out for now so we can investigate it in isolation...

**Pull request merged**